libraries/standard/mqtt/src/mqtt_state.c

```diff
@@ -0,0 +1,588 @@
+            break;
+    }
+    return isOutgoing;
+}
+
+static size_t _findInRecord( const MQTTPubAckInfo_t * records,
+                             size_t recordCount,
+                             uint16_t packetId,
+                             MQTTQoS_t * pQos,
+                             MQTTPublishState_t * pCurrentState )
+{
+    size_t index = 0;
+    *pCurrentState = MQTTStateNull;
+    if( packetId == MQTT_PACKET_ID_INVALID )
+    {
+        index = recordCount;
+    }
+    else
+    {
+        for( index = 0; index < recordCount; index++ )
+        {
+            if( records[ index ].packetId == packetId )
+            {
+                *pQos = records[ index ].qos;
+                *pCurrentState = records[ index ].publishState;
+                break;
+            }
+        }
+    }
+    return index;
+}
+
+static MQTTStatus_t _addRecord( MQTTPubAckInfo_t * records,
+                                size_t recordCount,
+                                uint16_t packetId,
+                                MQTTQoS_t qos,
+                                MQTTPublishState_t publishState )
+{
+    MQTTStatus_t status = MQTTNoMemory;
+    int32_t index = 0;
+    size_t availableIndex = recordCount;
+
+    assert( packetId != MQTT_PACKET_ID_INVALID );
+    assert( qos != MQTTQoS0 );
+
+    /* Start from end so first available index will be populated. */
+    for( index = ( (int32_t ) recordCount - 1 ); index >= 0; index-- )
+    {
+        if( records[ index ].packetId == MQTT_PACKET_ID_INVALID )
+        {
+            availableIndex = ( size_t ) index;
+        }
+        else if( records[ index ].packetId == packetId )
+        {
+            /* Collision. */
+            status = MQTTStateCollision;
+            availableIndex = recordCount;
+            break;
+        }
+        else
+        {
+            /* Empty else clause. */
+        }   
+    }
+
+    if( availableIndex < recordCount )
+    {
+        records[ availableIndex ].packetId = packetId;
+        records[ availableIndex ].qos = qos;
+        records[ availableIndex ].publishState = publishState;
+        status = MQTTSuccess;
+    }
+
+    return status;
+}
+
+static void _updateRecord( MQTTPubAckInfo_t * records,
+                           size_t recordCount,
+                           size_t recordIndex,
+                           MQTTPublishState_t newState,
+                           bool shouldDelete )
+{
+    assert( recordIndex < recordCount );
+
+    if( shouldDelete )
+    {
+        records[ recordIndex ].packetId = MQTT_PACKET_ID_INVALID;
+        records[ recordIndex ].qos = MQTTQoS0;
+        records[ recordIndex ].publishState = MQTTStateNull;
+    }
+    else
+    {
+        records[ recordIndex ].publishState = newState;
+    }
+
+}
+
+MQTTPublishState_t MQTT_ReserveState( MQTTContext_t * pMqttContext,
+                                      uint16_t packetId,
+                                      MQTTQoS_t qos )
```

Comment:
I realize there's no check for QoS 0. I will update this function for that case.